app/assets/javascripts/discourse/app/lib/highlight-syntax.js

```diff
@@ -1,35 +1,154 @@
-/*global hljs:true */
-let _moreLanguages = [];
-
+import { Promise } from "rsvp";
+import { getURLWithCDN } from "discourse-common/lib/get-url";
+import { next } from "@ember/runloop";
 import loadScript from "discourse/lib/load-script";
+import { isTesting } from "discourse-common/config/environment";
 
-export default function highlightSyntax($elem) {
-  const selector = Discourse.SiteSettings.autohighlight_all_code
-      ? "pre code"
-      : "pre code[class]",
-    path = Discourse.HighlightJSPath;
+const _moreLanguages = [];
+let _worker = null;
+let _workerPromise = null;
+const _pendingResolution = {};
+let _counter = 0;
+let _cachedResultsMap = new Map();
 
-  if (!path) {
-    return;
-  }
+const CACHE_SIZE = 100;
 
-  $(selector, $elem).each(function(i, e) {
-    $(e).removeClass("lang-auto");
-    loadScript(path).then(() => {
-      customHighlightJSLanguages();
-      hljs.highlightBlock(e);
+export function registerHighlightJSLanguage(name, fn) {
+  _moreLanguages.push({ name: name, fn: fn });
+}
+
+export default function highlightSyntax(elem) {
+  const selector = Discourse.SiteSettings.autohighlight_all_code
+    ? "pre code"
+    : "pre code[class]";
+
+  elem.querySelectorAll(selector).forEach(function(e) {
+    e.classList.remove("lang-auto");
+    let lang = null;
+    e.classList.forEach(c => {
+      if (c.startsWith("lang-")) {
+        lang = c.slice("lang-".length);
+      }
     });
+
+    asyncHighlight(e.textContent, lang).then(res =>
+      // Batch DOM changes using the Ember runloop
+      next(() => {
+        e.innerHTML = res;
+        e.classList.add("hljs");
+      })
+    );
   });
 }
 
-export function registerHighlightJSLanguage(name, fn) {
-  _moreLanguages.push({ name: name, fn: fn });
+function getWorker() {
+  if (_worker) return Promise.resolve(_worker);
+  if (_workerPromise) return _workerPromise;
+
+  const w = new Worker(Discourse.HighlightJSWorkerURL);
```

Comment:
Would attaching it to `Site.current()` be ok? Happy to update some others at the same time. At a quick glance, we have these similar variables:

- LetterAvatarVersion
- MarkdownItURL
- ServiceWorkerURL
- HighlightJSPath
- SvgSpritePath

set here:

https://github.com/discourse/discourse/blob/da0fc0a9d3889047b019fa90b6ed2ad06e89656e/app/assets/javascripts/discourse/app/pre-initializers/discourse-bootstrap.js#L63-L86

Reply:
`Site` should probably be fine, although again I'd advice to pass the site in and not call `current()` if possible. Still, as long as it's not on `Discourse` that works better.

As for the others I'd prefer to approach them on a case by case basis. It's probable some would be better off elsewhere.

Reply:
Trying to work out the cleanest way to design this. @eviltrout how do you feel about having an Ember Service for syntax highlighting? 

Then rather than importing a function, consumers would `inject` the highlighting service and call a method on it. That way, the service will have direct access to site settings and the singleton site via dependency injection.

Reply:
That sounds quite nice! Let's do it.